parsers/CA_NS.py

```diff
@@ -18,35 +44,73 @@ def _get_ns_info(requests_obj):
     production = []
     imports = []
     for mix in mix_data:
-        corresponding_load = [load_period for load_period in load_data
-                              if load_period['datetime'] == mix['datetime']]
+        percent_mix = {
+            'coal': mix['Solid Fuel'] / 100.0,
+            'gas': (mix['HFO/Natural Gas'] + mix['CT\'s'] + mix['LM 6000\'s']) / 100.0,
+            'biomass': mix['Biomass'] / 100.0,
+            'hydro': mix['Hydro'] / 100.0,
+            'wind': mix['Wind'] / 100.0,
+            'imports': mix['Imports'] / 100.0
+        }
+
+        # datetime is in format '/Date(1493924400000)/'
+        # get the timestamp 1493924400 (cutting out last three zeros as well)
+        data_timestamp = int(mix['datetime'][6:-5])
+        data_date = arrow.get(data_timestamp).datetime
+
+        # validate
+        valid = True
+        for gen_type, value in percent_mix.items():
+            percent_bounds = valid_percent[gen_type]
+            if not (percent_bounds[0] <= value <= percent_bounds[1]):
+                # skip this datapoint in the loop
+                valid = False
+                logger.warning(
+                    'CA-NS: discarding datapoint at {dt} '
+                    'due to {fuel} value out of bounds: {value}'.format(
+                        dt=data_date, fuel=gen_type, value=value))
+        if not valid:
+            # continue the outer loop, not the inner
+            continue
 
         # in mix_data, the values are expressed as percentages,
-        # e.g. "Solid Fuel":  45.76 meaning 45.76%
-        # Divide load by 100.0 to simplify calculations later on.
+        # and have to be multiplied by load to find the actual MW value.
+        corresponding_load = [load_period for load_period in load_data
+                              if load_period['datetime'] == mix['datetime']]
         if corresponding_load:
-            load = corresponding_load[0]['Base Load'] / 100.0
+            load = corresponding_load[0]['Base Load']
         else:
             # if not found, assume 1244 MW, based on average yearly electricity available for use
             # in 2014 and 2015 (Statistics Canada table Table 127-0008 for Nova Scotia)
-            load = 1244 / 100.0
-
-        # datetime is in format '/Date(1493924400000)/'
-        # get the timestamp 1493924400 (cutting out last three zeros as well)
-        data_timestamp = int(mix['datetime'][6:-5])
-        data_date = arrow.get(data_timestamp).datetime
+            load = 1244
+            logger.warning('CA-NS: unable to find load for {}, assuming 1244 MW'.format(data_date))
```

Comment:
can you add `extra={'key': 'CA-NS'}` as extra argument to make sure the zone code gets registered?

Reply:
Sure - how should it look in total? Is there an existing example? Like this?

```
logger.warning(
    'CA-NS: unable to find load for {}, assuming 1244 MW'.format(data_date),
    extra={'key':'CA-NS'})
```

And do that in all warnings I log?

Reply:
exactly. You can even drop the `CA-NS: ` prefix from the message.